dcrec/secp256k1/ecdsa/signature.go

```diff
@@ -581,82 +703,14 @@ func signRFC6979(privKey *secp256k1.PrivateKey, hash []byte) (*Signature, byte)
 		// private key, message being signed, and iteration count.
 		k := secp256k1.NonceRFC6979(privKeyBytes[:], hash, nil, nil, iteration)
 
-		// Step 2.
-		//
-		// Compute kG
-		//
-		// Note that the point must be in affine coordinates.
-		var kG secp256k1.JacobianPoint
-		secp256k1.ScalarBaseMultNonConst(k, &kG)
-		kG.ToAffine()
-
-		// Step 3.
-		//
-		// r = kG.x mod N
-		// Repeat from step 1 if r = 0
-		r, overflow := fieldToModNScalar(&kG.X)
-		if r.IsZero() {
-			k.Zero()
-			continue
-		}
-
-		// Since the secp256k1 curve has a cofactor of 1, when recovering a
-		// public key from an ECDSA signature over it, there are four possible
-		// candidates corresponding to the following cases:
-		//
-		// 1) The X coord of the random point is < N and its Y coord even
-		// 2) The X coord of the random point is < N and its Y coord is odd
-		// 3) The X coord of the random point is >= N and its Y coord is even
-		// 4) The X coord of the random point is >= N and its Y coord is odd
-		//
-		// Rather than forcing the recovery procedure to check all possible
-		// cases, this creates a recovery code that uniquely identifies which of
-		// the cases apply by making use of 2 bits.  Bit 0 identifies the
-		// oddness case and Bit 1 identifies the overflow case (aka when the X
-		// coord >= N).
-		//
-		// It is also worth noting that making use of Hasse's theorem shows
-		// there are around log_2((p-n)/p) ~= -127.65 ~= 1 in 2^127 points where
-		// the X coordinate is >= N.  It is not possible to calculate these
-		// points since that would require breaking the ECDLP, but, in practice
-		// this strongly implies with extremely high probability that there are
-		// only a few actual points for which this case is true.
-		pubKeyRecoveryCode := byte(overflow<<1) | byte(kG.Y.IsOddBit())
-
-		// Step 4.
-		//
-		// e = H(m)
-		//
-		// Note that this actually sets e = H(m) mod N which is correct since
-		// it is only used in step 5 which itself is mod N.
-		var e secp256k1.ModNScalar
-		e.SetByteSlice(hash)
-
-		// Step 5 with modification B.
-		//
-		// s = k^-1(e + dr) mod N
-		// Repeat from step 1 if s = 0
-		// s = -s if s > N/2
-		kInv := new(secp256k1.ModNScalar).InverseValNonConst(k)
+		// Steps 2-6.
+		sig, pubKeyRecoveryCode, success := sign(privKeyScalar, k, hash)
 		k.Zero()
-		s := new(secp256k1.ModNScalar).Mul2(privKeyScalar, &r).Add(&e).Mul(kInv)
-		if s.IsZero() {
+		if !success {
 			continue
```

Comment:
If the sign() function generates an `r` or `s` with a value of zero, that function yields "success = false".
According to the algorithm notes, you must start over from step 1 (a new nonce) in this case.
By continuing on to the next iteration of this for loop, the iteration is incremented, a new nonce is calculated, and it tries the sign() function again.

Just want to make sure I understand the iteration for loop correctly. Seems like each individual iteration isn't important, and this will use as many iterations as required (though probably only 1 attempt) to generate a valid sig & recovery code. Correct?

This could all be apparent already, but a comment here might make it a bit more clear:
> // Retry from step 1 (using a new nonce)

Reply:
Yes, that's correct.  Maybe it's not quite as obvious from the diff due to being interspersed with the moved code, but it seems pretty apparent from the final code to me since it's only 4 lines of code away from the iteration loop and the comment on the nonce call says `Generate a deterministic nonce in [1, N-1] parameterized by the private key, message being signed, and iteration count.`

I can add it if you really feel it's non-obvious, but I suspect it would be redundant in the final code due to its proximity.

Reply:
Nope, I initially didn't realize that the entire point of the for loop is to keep trying nonces until one works. Makes more sense now. I think it just took me a while to understand, and will definitely be more clear outside of the diff.